Accessibility: Add text labels to toolbar buttons

JohJakob opened the pull request:
## Changes

- Adds text labels to the toolbar buttons in the projects view and in the graph editor (macOS + iPadOS)
- Changes the label of the “Find Node” button to “Go to Node”
- Always displays the “Go Up” toolbar button (disabled if the user cannot go up any further)
- Removes the rotation effect of the “Restart Prototype” button’s icon (did not work anymore with the simplified button structure and should be replaced with a symbol effect in the future)

![Toolbar Buttons iPad](https://github.com/user-attachments/assets/41456d32-e318-4680-a50c-d449ea202f99)

![Toolbar Buttons Mac Projects](https://github.com/user-attachments/assets/ad60fc1a-4bea-449c-969c-f6d39b8ff7ef)

![Toolbar Buttons Mac Projects Text Labels](https://github.com/user-attachments/assets/5a053443-f4d6-405e-962c-a39c11befa5e)

![Toolbar Buttons Mac](https://github.com/user-attachments/assets/855eeaf5-dcb2-4ae2-867a-79d163fe9d9c)

![Toolbar Buttons Mac Text Labels](https://github.com/user-attachments/assets/559881f5-af2d-4a77-870c-ec39e76bea02)

Closes #1221

## Discussion

As you can see, the text labels are not really optimal. On macOS, they are way too subtle (looks like the disabled state), but I was unable to change that. Creating the buttons using the hacky `Menu` trick is also probably not sustainable. A proper `NSToolbar` would be a better solution for the future, I suppose. This would also allow for getting rid of the text labels beneath the title in the projects view and the document title in the graph editor.

I was also unable to make the text labels change based on state (but it works for the icons?) so I chose state-independent labels like “Toggle Preview.”

JohJakob committed:
Accessibility: Add text labels to graph toolbar buttons

- Adds text labels and tooltips to the toolbar buttons in the graph editor (macOS)

JohJakob committed:
Accessibility: Add text labels to projects view toolbar buttons

- Adds text labels and tooltips to the toolbar buttons in the projects view (macOS)

JohJakob committed:
Accessibility: Add text labels to toolbar buttons (iPadOS)

- Adds text labels to the toolbar buttons in the projects view and in the graph editor (iPadOS)
- Removes code for rotating the “Restart Prototype” button’s icon (replace with a symbol effect in the future)

JohJakob committed:
Chore: Make label consistent

- Updates the label for the “Go to Node” toolbar button on iPadOS for consistency with macOS

ellbosch commented:
@JohJakob thanks a ton for putting something together here! I just synced with the team, I think we can proceed forward with these changes if we do tooltips instead of permanent text labels. Hopefully this is still good accessible design, or at least an improvement. We see other apps doing this, like so:

<img width="233" alt="CleanShot 2025-05-09 at 13 24 55@2x" src="https://github.com/user-attachments/assets/4a68c09a-605e-4417-b643-87e94c76ee15" />

Otherwise my bet is you're running into some of the many Catalyst roadblocks we've hit. Let us know your thoughts here and if we can help in any way. HUGE thank you this PR ❤️

ellbosch commented on Stitch/Graph/Toolbar/View/TopBarButtonView.swift:
```diff
@@ -107,26 +105,33 @@ struct iPadGraphTopBarButtons: View {
-            
+            iPadNavBarButton(title: "Go Up", iconName: .sfSymbol(.GO_UP_ONE_TRAVERSAL_LEVEL_SF_SYMBOL_NAME), tooltip: "") { dispatch(GoUpOneTraversalLevel())
+            }
+            .disabled(!hasActiveGroupFocused)
```
I think the preference here is to actually hide the group traversal button if at the root hierarchy—I forget if `.disabled` just disables a visible button.

JohJakob replied:
Yes, `.disabled()` keeps the button visible. I prefer that to hiding the button, but that’s a personal preference. I think it is generally better to display all _potential_ actions and not hide and show interactive elements conditionally.

I’m open to changing it back to the previous behaviour (with the addition of making it invisible to assistive technology as well since setting the opacity to 0 still keeps it in the accessibility tree).

ellbosch replied:
Let's leave your changes in for now! That'll allow us to get a feel for the change here. I'm definitely not against it.

ellbosch commented on Stitch/Graph/Toolbar/View/TopBarButtonView.swift:
```diff
@@ -34,27 +34,25 @@ struct iPadTopBarButtonStyle: ViewModifier {
         } primaryAction: {
             action()
         }
-        .rotation3DEffect(Angle(degrees: rotationZ),
```
I believe this rotation call impacts prototype restart icon on iPad—doesn't work on Catalyst. However let me know if removing this functionality enables better accessible design here.

JohJakob replied:
This removal doesn’t affect accessibility so perhaps it’s a bit out of scope here. I guess the wonky rotation just bothered me because it doesn’t rotate around the centre of the circle due to the arrowhead. In any case, using the rotation symbol effect would be a better solution.

ellbosch replied:
Haha yes totally agree it's a little wonky. I'd rather have some rotation than none, let's leave for now but definitely not against better animations here.

ellbosch replied:
Got it. In that case let's leave it commented out and add a new comment explaining why it's disabled.

JohJakob commented:
@ellbosch The text labels are not permanent. The user can choose between three toolbar display modes by doing a secondary click on the toolbar: _Icon and Text_, _Icon Only_, or _Text Only_. This is also the case in other apps like Mail. So this change adds the missing text labels for the actions that didn’t have a label before. Oh, and I forgot to mention that I also added tooltips (optional; if not provided, the label title will be shown as a tooltip).

ellbosch commented:
> @ellbosch The text labels are not permanent. The user can choose between three toolbar display modes by doing a secondary click on the toolbar: _Icon and Text_, _Icon Only_, or _Text Only_

Oooooh nice!! This makes a lot of sense. We'll take a closer look at this PR and get back here with more feedback.

ellbosch commented:
I just pulled the branch and explored things more. It's honestly great. Some things to sort out first and then we can merge:

* iPadOS isn’t compiling—I can help out here if too difficult for you to test iPad, just let me know.
* Any chance we can remove the label under the project title? If not, let’s left-align it and replace it with “Document Name”.
  ![image](https://github.com/user-attachments/assets/a73f8a1c-f33a-4eef-a2db-e34cd6c8906f)

* Same issue with home screen “Stitch Projects”, let's see if we can remove this.
  <img width="145" alt="image 2" src="https://github.com/user-attachments/assets/77a87d28-7683-46e9-b515-a8b4dd669f7f" />

* Shorter label names to better utilize the space and prevent overflow likelihood:
  * Restart Prototype -> Restart
  * Toggle Fullscreen Preview -> Fullscreen
  * Contact Stitch -> Contact
  * Project Settings -> Settings

JohJakob committed:
Accessibility: Shorten toolbar button labels

JohJakob committed:
Chore: Replace empty text view with spacer

JohJakob committed:
Chore: Use `.topBarLeading` item placement

- Replaces the deprecated `.navigationBarLeading` toolbar item placement with `.topBarLeading`

JohJakob committed:
Accessibility: Use button for toggling project title modal

- Replaces the project title text view with the attached `.onTapGesture()` modifier with a button for improved accessibility
- Replaces the two document events `CatalystProjectTitleModalOpened` and `CatalystProjectTitleModalClosed` with a single `CatalystProjectTitleModalToggled` event

JohJakob committed:
Chore: Correct indentation

JohJakob committed:
Merge branch 'development' into fix/toolbar-button-labels

ellbosch commented on Stitch/Graph/Toolbar/View/CatalystNavigationBarHelperViews.swift:
```diff
@@ -79,14 +74,16 @@ struct CatalystNavBarProjectTitleDisplayView: View {
+                .padding(6)
+                .frame(width: 260, height: 16, alignment: .leading)
+        }
+        .buttonStyle(.plain)
```
If this `buttonStyle` is a new change we need to make sure things look ok on both iPad and Catalyst. Goal here being to keep the same design.

ellbosch replied:
nvm we're looking good here.

ellbosch commented:
Alrighty @JohJakob we're real close to merging here. I just made a quick commit at `d15201f7487dbfa7bec9546ee2e7a22c25800ddb` to resolve the compiler issues... hopefully that works as a cherry-pick, if not there's also my `u/ellbosch/2505/accessibility` branch. Also apologies here as you stumbled into conflicts with our recent changes to the toolbar.

Please give us a ping once the conflicts are resolved!! Thanks so much for your help 😍

JohJakob committed:
Chore: Add icon rotation back in

- Adds the previously removed code for rotating icons in the toolbar
- Disables the icon rotation modifier

JohJakob committed:
Fix: Updates AI Generation/Correction toolbar button

JohJakob committed:
Refactor: Make parameter name clearer

JohJakob committed:
Cleanup: Replace `CatalystHomescreenNavBarButton` with `CatalystNavBarButton`

JohJakob committed:
Merge branch 'development' into fix/toolbar-button-labels

JohJakob committed:
Fix: Update view calls

- Updates view calls

JohJakob commented:
I fixed the merge conflicts. Hopefully, everything is there now. The toolbars look complete to me at least. Could you please check that I didn’t miss anything important, @ellbosch?

That’s how it looks on Mac in the “Icon and Text” display mode (the Sidebar button gets pushed into the overflow menu if I’m not in fullscreen for some reason):

![Projects View Toolbar Mac](https://github.com/user-attachments/assets/6c2e262b-0f54-49e6-b4d3-401f5aa4047e)
![Graph Editor Toolbar Mac](https://github.com/user-attachments/assets/4b07b357-9489-4115-a97a-3996e8cf9655)

And on iPad:

![Projects View Toolbar iPad](https://github.com/user-attachments/assets/b490cfa5-2f80-451a-a0c6-dda39bfb3223)
![Graph Editor Toolbar iPad](https://github.com/user-attachments/assets/d32e3c3f-932d-4dc4-bac4-6f49a534cabe)
![Graph Editor Context Menu iPad](https://github.com/user-attachments/assets/88030e0d-3efe-4d88-956e-323b4f9f2997)
